### PR TITLE
[16.0][MIG] pricelist_brand

### PR DESCRIPTION
**Pull request by david-s73**
Migration to 16, reference: https://github.com/odoo/odoo/pull/79605

### CODE REVIEW

File: pricelist_brand/models/product_pricelist.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+from odoo.osv import expression
+
+
+class ProductPricelist(models.Model):
+
+    _inherit = "product.pricelist"
+
+    def _get_applicable_rules_domain(self, products, date, **kwargs):
+        res = super()._get_applicable_rules_domain(products, date)
+        print(res)
```
Comment (ferran-S73) [inline pricelist_brand/models/product_pricelist.py]:
```suggestion
```

**Review by ferran-S73 — APPROVED** [review]

**Comment by bosd** [comment]
@david-s73 Can you please fix pre-commit?

**Review by sbejaoui — APPROVED** [review]
Code review

**Review by bosd — CHANGES_REQUESTED** [review]
The commit history is missing. Can you please follow the migration guide?

**Review by Alexgars73 — APPROVED** [review]

**Comment by david-s73** [comment]
@bosd I can see that all previous commits have been carried over in the migration. Could you clarify what you mean?

**Review by Reyes4711-S73 — APPROVED** [review]
LGTM

**Comment by Reyes4711-S73** [comment]
@sbejaoui Please, can you merge this PR?

**Comment by bosd** [comment]
> @bosd I can see that all previous commits have been carried over in the migration. Could you clarify what you mean?

i was expecting more of a commit history. But it seems that there is none for this module. so it is ok now.

**Review by bosd — APPROVED** [review]

**Comment by Reyes4711-S73** [comment]
@OCA/brand-maintainers Please, can you review/merge this PR?

**Comment by andreampiovesana** [comment]
merge?

**Comment by andreampiovesana** [comment]
?

**Comment by bosd** [comment]
> merge?

@max3903 Can you :pray:

**Review by marcos-mendez — CHANGES_REQUESTED** [review]
**Automated Review -- Tests Failed**

### 1. **Root cause** of the test failure

The test failure is caused by a missing `brand_id` field in the pricelist item when the `applied_on` field is set to `'0_product_brand'`. The code expects this field to be present and populated, but it is not being set during the test creation, leading to an unexpected behavior or error during pricelist application.

### 2. **Suggested fix**

In `pricelist_brand/models/product_pricelist.py`, ensure that when `applied_on` is set to `'0_product_brand'`, the `brand_id` field is properly required and validated. Specifically, add a domain or constraint to enforce that `brand_id` is set when `applied_on` is `'0_product_brand'`.

Additionally, in the test file `pricelist_brand/tests/test_product_pricelist.py`, ensure that when creating a pricelist item with `applied_on='0_product_brand'`, the `brand_id` is explicitly set.

### 3. **Additional code issues**

- **Missing validation**: There is no validation to ensure that `brand_id` is required when `applied_on` is `'0_product_brand'`. This could lead to inconsistent data or runtime errors.
- **Test coverage gap**: The test does not validate that pricelist items with `applied_on='0_product_brand'` are correctly applied only to products of the specified brand.

### 4. **Test improvements**

Add the following test cases to `test_product_pricelist.py` using `SavepointCase` for better isolation:

```python
def test_pricelist_item_brand_applied(self):
    # Create a brand
    brand = self.env['product.brand'].create({'name': 'Test Brand'})
    
    # Create a product with the brand
    product = self.env['product.product'].create({
        'name': 'Test Product',
        'brand_id': brand.id,
    })
    
    # Create a pricelist item with brand applied
    pricelist_item = self.env['product.pricelist.item'].create({
        'pricelist_id': self.pricelist.id,
        'applied_on': '0_product_brand',
        'brand_id': brand.id,
        'fixed_price': 100.0,
    })
    
    # Check that the price is applied correctly
    price = self.pricelist.get_product_price(product, 1.0)
    self.assertEqual(price, 100.0)

def test_pricelist_item_brand_not_applied(self):
    # Create two brands
    brand1 = self.env['product.brand'].create({'name': 'Brand 1'})
    brand2 = self.env['product.brand'].create({'name': 'Brand 2'})
    
    # Create a product with brand1
    product = self.env['product.product'].create({
        'name': 'Test Product',
        'brand_id': brand1.id,
    })
    
    # Create a pricelist item with brand2 applied
    pricelist_item = self.env['product.pricelist.item'].create({
        'pricelist_id': self.pricelist.id,
        'applied_on': '0_product_brand',
        'brand_id': brand2.id,
        'fixed_price': 100.0,
    })
    
    # Check that the price is NOT applied (should be default price)
    price = self.pricelist.get_product_price(product, 1.0)
    self.assertNotEqual(price, 100.0)
```

Use `SavepointCase` for these tests to ensure that each test runs in isolation and does not affect other tests. This follows OCA testing patterns for better reliability and maintainability.

---
_Environment: Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by OCA PR Reviewer + qwen3-coder:30b_

**Review by marcos-mendez — APPROVED** [review]
**Automated Review -- Tests Passed**

All tests for `pricelist_brand` passed successfully on Odoo 16.0.
Environment: Minikube + K8s Job + oca-ci/py3.10-odoo16.0

---
### Test Coverage Suggestions

## Coverage Gaps

The existing tests cover basic functionality, but there are some gaps:

1. **Domain filtering logic** in `_get_applicable_rules_domain` is not fully tested for edge cases (e.g., empty brands, multiple brands)
2. **Constraint validation** for `product_brand_id` is tested, but not the case where `applied_on` is changed from "Brand" to another option
3. **Name and price computation** logic is tested, but not thoroughly for all combinations of brand + other applied_on values
4. **Create/write consistency logic** in `create` and `write` methods lacks tests for:
   - Changing `applied_on` from "Brand" to other values
   - Ensuring that conflicting fields are cleared correctly

## Suggested Test Cases

```python
def test_pricelist_item_brand_domain_filtering(self):
    """Test that pricelist rules are correctly filtered by brand in domain."""
    # Create products with different brands
    brand1 = self.env['product.brand'].create({'name': 'Brand1'})
    brand2 = self.env['product.brand'].create({'name': 'Brand2'})
    product1 = self.env['product.template'].create({
        'name': 'Product1', 'product_brand_id': brand1.id
    })
    product2 = self.env['product.template'].create({
        'name': 'Product2', 'product_brand_id': brand2.id
    })
    pricelist = self.env['product.pricelist'].create({'name': 'Test Pricelist'})
    item = self.env['product.pricelist.item'].create({
        'pricelist_id': pricelist.id,
        'applied_on': '25_brand',
        'product_brand_id': brand1.id,
        'fixed_price': 100.0,
    })
    # Ensure rule applies only to brand1 products
    domain = pricelist._get_applicable_rules_domain(product1, False)
    self.assertIn(item.id, self.env['product.pricelist.item'].search(domain).ids)
    domain = pricelist._get_applicable_rules_domain(product2, False)
    self.assertNotIn(item.id, self.env['product.pricelist.item'].search(domain).ids)

def test_pricelist_item_brand_constraint_on_write(self):
    """Test that constraint is enforced when changing applied_on to 'Brand'."""
    pricelist = self.env['product.pricelist'].create({'name': 'Test Pricelist'})
    item = self.env['product.pricelist.item'].create({
        'pricelist_id': pricelist.id,
        'applied_on': '3_global',
        'fixed_price': 100.0,
    })
    # Change to brand rule without specifying brand - should raise ValidationError
    with self.assertRaises(ValidationError):
        item.write({'applied_on': '25_brand'})

def test_pricelist_item_brand_create_consistency(self):
    """Test that creating a brand rule clears conflicting fields."""
    pricelist = self.env['product.pricelist'].create({'name': 'Test Pricelist'})
    item = self.env['product.pricelist.item'].create({
        'pricelist_id': pricelist.id,
        'applied_on': '25_brand',
        'product_brand_id': self.env['product.brand'].create({'name': 'Test'}).id,
        'product_id': self.env['product.product'].create({'name': 'Test'}).id,  # Should be cleared
        'fixed_price': 100.0,
    })
    self.assertFalse(item.product_id)
    self.assertFalse(item.product_tmpl_id)
    self.assertFalse(item.categ_id)
```

## Codecov Risk

The following methods have high risk of reduced coverage if not tested:
- `_get_applicable_rules_domain` - domain construction logic
- `create` and `write` methods in `ProductPricelistItem` - field clearing logic
- `_compute_name_and_price` - name formatting logic

These methods involve complex conditional logic and field interactions that are not fully covered by current tests.

---
### ⚠️ PR Aging Alert: CRITICAL

This PR by @david-s73 has been **waiting for 769 days** — that is over **25 months** without being merged or closed.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! Thanks for your contribution to OCA. I reviewed and approved this PR. If any of you have a moment, I would really appreciate a review on my open PR(s):

**My open PRs across OCA:**
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward faster. Thank you!

---
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

**Comment by Reyes4711-S73** [comment]
@OCA/brand-maintainers Please, can you review/merge this PR?

**Comment by pcastelovigo** [comment]
@rvalyi this was blocked too because copier template, can be merged now

**Review by BhaveshHeliconia — APPROVED** [review]
Code and functional review LGTM!

**Review by pcastelovigo — APPROVED** [review]

**Review by bofiltd — APPROVED** [review]
LGTM 👍

**Comment by Reyes4711-S73** [comment]
@sbejaoui Please I made a rebase.
Can you retry again?

**Comment by pcastelovigo** [comment]
this PR is cursed

**Comment by bosd** [comment]
> this PR is cursed

We just need 1 more review :pray:

**Review by pablo-cort-s73 — APPROVED** [review]
LGTM

**Review by marcos-mendez — APPROVED** [review]
OK
